crates/red_knot_server/src/system.rs

```diff
@@ -0,0 +1,243 @@
+        let index = self.index();
+        let key = index.key_from_url(url);
+        index.make_document_ref(key).ok_or_else(|| {
+            std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                "Document not found in the index",
+            )
+        })
+    }
+
+    fn system_path_to_document_ref(&self, path: &SystemPath) -> Result<DocumentQuery> {
+        let url = Url::from_file_path(path.as_std_path()).map_err(|()| {
+            std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                format!("Failed to convert system path to URL: {path:?}"),
+            )
+        })?;
+        self.make_document_ref(url)
+    }
+
+    fn system_virtual_path_to_document_ref(
+        &self,
+        path: &SystemVirtualPath,
+    ) -> Result<DocumentQuery> {
+        let url = Url::parse(path.as_str()).map_err(|_| {
+            std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                format!("Failed to convert virtual path to URL: {path:?}"),
+            )
+        })?;
+        self.make_document_ref(url)
+    }
+}
+
+impl System for LSPSystem {
+    fn path_metadata(&self, path: &SystemPath) -> Result<Metadata> {
+        let document = self.system_path_to_document_ref(path);
+
+        // First, we need to check if the document is opened in the editor. If it is, we need to
+        // use the document's version as the file revision. Otherwise, fall back to the OS system.
+        match document {
+            Ok(document) => {
+                // The file revision is just an opaque number which doesn't have any significant
+                // meaning other than that the file has changed if the revisions are different.
+                #[allow(clippy::cast_sign_loss)]
+                Ok(Metadata::new(
+                    FileRevision::new(document.version() as u128),
+                    None,
+                    FileType::File,
+                ))
+            }
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                self.os_system.path_metadata(path)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    fn canonicalize_path(&self, path: &SystemPath) -> Result<SystemPathBuf> {
+        self.os_system.canonicalize_path(path)
+    }
+
+    fn read_to_string(&self, path: &SystemPath) -> Result<String> {
+        let document = self.system_path_to_document_ref(path);
+
+        match document {
+            Ok(document) => {
+                if let DocumentQuery::Text { document, .. } = &document {
+                    Ok(document.contents().to_string())
+                } else {
+                    Err(not_a_text_document(path))
+                }
+            }
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                self.os_system.read_to_string(path)
+            }
+            Err(err) => Err(err),
+        }
```

Comment:
I'm not sure how this could work because `system_path_to_document_ref` could give two errors, (1) invalid input because system path to url conversion failed or (2) no document exists for the url. We should only fall back to the OS system in case it's (2).

If we convert `system_path_to_document_ref` to return an `Option`, then we can't tell the difference between the two.

Reply:
Sorry. I missed the second \` in the first sentence and it resulted in the sentence being truncated. I would return a `Result<Option<Document>>`

Reply:
Oh, I see. Yeah, that could work.